File: 3634-minimum-removals-to-balance-array/3634-minimum-removals-to-balance-array.go
func minRemoval(nums []int, k int) int {
    if len(nums) == 1 { return 0 }

    sort.Slice(nums, func(i, j int) bool { return nums[i] < nums[j] })
    
    left := 0
    maxWindow := 1
    for right := 1; right < len(nums); right++ {
        for nums[left] * k < nums[right] {
            left++
        }

        if nums[right] <= nums[left] * k {
            maxWindow = max(maxWindow, right - left + 1)
        }
    }

    return len(nums) - maxWindow
}

// array, senior(?), sliding window, sorting
// time: O(nlog(n))
// space: O(1)

File: 3634-minimum-removals-to-balance-array/3634-minimum-removals-to-balance-array.py
class Solution:
    def minRemoval(self, nums: List[int], k: int) -> int:
        if len(nums) == 1: return 0

        nums.sort()

        left = 0
        max_window = 1
        for right in range(1, len(nums)):
            while nums[right] > nums[left] * k:
                left += 1
            
            if nums[right] <= nums[left] * k:
                max_window = max(max_window, right - left + 1)
        
        return len(nums) - max_window

# array, sliding window, sorting
# time: O(nlog(n))
# space: O(1)